### PR TITLE
feat(table): add row when predicate

### LINKED ISSUE #5887: [Table] Apply conditional rows using by a when-predicate
Add the ability to define multiple `<cdk-row>` directives which can have unique sets of columns and event/property bindings.

```
<md-row *cdkRowDef="let row; columns: usersColumns;"></md-row>
<md-row *cdkRowDef="let row; columns: usersColumns; when:isSpecialUser" (hover)="showFireworks()"></md-row>
<md-row *cdkRowDef="let row; columns: adminsColumns; when:isAdmin" (click)="selectAdmin(row)"></md-row>
```

### PR DESCRIPTION
**Pull request by andrewseguin**
Fixes #5887

### CODE REVIEW

File: src/lib/table/table.spec.ts [hunk 1]
```diff
@@ -7,17 +7,17 @@ import {MdTableModule} from './index';
 import {MdTable} from './table';
 
 describe('MdTable', () => {
-  let fixture: ComponentFixture<SimpleMdTableApp>;
+  let fixture: ComponentFixture<MdTableApp>;
```
Comment (jelbourn) [inline src/lib/table/table.spec.ts]: * Add test case for `when` missing on multiple rows?
* Add test case for using the default row def?

There's also no test now for a table that just has one row def

File: src/cdk/table/table.ts [hunk 2]
```diff
@@ -300,14 +303,28 @@ export class CdkTable<T> implements CollectionViewer {
...
   /**
+   * Finds the matching row definition that should be used for this row data. If there is only
+   * one row definition, it is returned. Otherwise, find the row definition that has a when
+   * predicate that returns true with the data. If none return true, return the default row
+   * definition.
+   */
+  _getRowDef(data: T, i: number): CdkRowDef<T> {
+    if (this._rowDefs.length == 1) { return this._rowDefs.first; }
+
+    let rowDef = this._rowDefs.find(def => def.when && def.when(data, i)) || this._defaultRowDef;
```
[the remark this answers is not in the document]
Comment (andrewseguin) [reply src/cdk/table/table.ts]: Would this mean everyone would need to include a when function when defining their rows? I figured we wanted to let them leave it off for a row and it'll be the default

File: src/lib/table/table.spec.ts (hunk 1 above)
Comment (andrewseguin) [reply src/lib/table/table.spec.ts]: Added test for missing when on multiple rows. Default row def is used for any case where there is no `when`, e.g. `SimpleCdkTableApp`

Note, majority of testing for the when logic is in `cdk/table.ts`. Goal of the `lib/table.ts` test is to show that the `MdRowDef` successfully reads the `when` input and passes it on to the `CdkRowDef`.

File: src/cdk/table/table.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jelbourn) [reply src/cdk/table/table.ts]: In that case would defaulting to `() => true` do the same thing? The first row with no `when` will be the default, all other rows with no `when` will be ignored.

[the remark this answers is not in the document]
Comment (andrewseguin) [reply src/cdk/table/table.ts]: Yup, that's fine with me. I think I was trying to catch the user's error ahead of time if they didn't provide a default. But it's not a perfect thing to catch anyways (all rows could have when but still catch 100% of cases)

[the remark this answers is not in the document]
Comment (andrewseguin) [reply src/cdk/table/table.ts]: Righto - I remember why its needed. If we default to true, that row is already our match when we look for a row def. If we default to `() => false` then it forces users to always define a when function

**Pull request merged**